### PR TITLE
Fix contiguity of expanded tensor

### PR DESCRIPTION
**Pull request by zasdfgbnm**
- Make `TensorViewBuilder` able to generate expanded tensor.
- Fixes a problem on the contiguity found while working on https://github.com/csarofeen/pytorch/pull/2044. Note that an expanded dimension and the dimension before can never be contiguous. Without fixing this, the added test will generate silent wrong result.
- Validate contiguity of expanded dimensions during lowering

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 1]
```diff
@@ -1310,27 +1313,53 @@ TensorViewBuilder& TensorViewBuilder::shape(std::vector<Val*> shape) {
...
 TensorView* TensorViewBuilder::build() const {
   // Build the domain
   std::vector<IterDomain*> domain(ndims_, nullptr);
   for (const auto i : c10::irange(ndims_)) {
+    bool is_expanded = false;
+    Val* extent = nullptr;
+    Val* expanded_extent = nullptr;
+    Val** shape_extent = &extent;
+    if (!expanded_.empty()) {
+      is_expanded = expanded_[i];
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Just in case, please use `expanded_.at(i)` to avoid silent memory errors

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 2]
```diff
@@ -1289,6 +1290,8 @@ TensorViewBuilder& TensorViewBuilder::shape(const std::vector<int64_t>& shape) {
   for (int64_t i : shape) {
     if (i == -1) {
       shape_.emplace_back(IrBuilder::create<Int>());
+    } else if (i == 1) {
+      shape_.emplace_back(FusionGuard::getCurFusion()->oneVal());
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:1294]: Is there any specific reason to use the one val? Do we also want to use the zero val?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 3]
```diff
@@ -1310,27 +1313,53 @@ TensorViewBuilder& TensorViewBuilder::shape(std::vector<Val*> shape) {
...
-        domain[i] = IterDomainBuilder(
-                        FusionGuard::getCurFusion()->zeroVal(),
-                        FusionGuard::getCurFusion()->oneVal())
-                        .iter_type(IterType::Broadcast)
-                        .build();
-      } else {
-        domain[i] =
-            IterDomainBuilder(FusionGuard::getCurFusion()->zeroVal(), shape_[i])
-                .build();
+      *shape_extent = shape_[i];
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp]: Same here

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 4]
```diff
@@ -1310,27 +1313,53 @@ TensorViewBuilder& TensorViewBuilder::shape(std::vector<Val*> shape) {
...
+}
+
 TensorView* TensorViewBuilder::build() const {
   // Build the domain
   std::vector<IterDomain*> domain(ndims_, nullptr);
   for (const auto i : c10::irange(ndims_)) {
+    bool is_expanded = false;
+    Val* extent = nullptr;
+    Val* expanded_extent = nullptr;
+    Val** shape_extent = &extent;
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:1342]: What is `shape_extent`? Why is this necessary?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp [hunk 5]
```diff
@@ -1310,27 +1313,53 @@ TensorViewBuilder& TensorViewBuilder::shape(std::vector<Val*> shape) {
...
-                        FusionGuard::getCurFusion()->oneVal())
-                        .iter_type(IterType::Broadcast)
-                        .build();
-      } else {
-        domain[i] =
-            IterDomainBuilder(FusionGuard::getCurFusion()->zeroVal(), shape_[i])
-                .build();
+      *shape_extent = shape_[i];
+    }
+    IterDomainBuilder builder(FusionGuard::getCurFusion()->zeroVal(), extent);
```
Comment (naoyam) [inline torch/csrc/jit/codegen/cuda/tensor_view.cpp:1356]: Should we do some sanity checks? If expanded is true, the extent should be 1, right?

**Comment by naoyam** [comment]
Contiguity and expanded broadcasts are still making me feel uneasy as I've yet got a clear idea of what they mean in PyTorch. Can you please explain why expanded domains and their left domains should never be contiguous?

**Comment by zasdfgbnm** [comment]
@naoyam For example, if you have a contiguous tensor of shape `(4, 5, 1, 6)`, then the stride could be `(30, 6, whatever, 1)`, but if you expand this tensor to `(4, 5, 10, 6)`, the stride has to be `(30, 6, 0, 1)`. And by definition, if we have `true` contiguity at the expanded dimension (dim 2), its stride must be `stride[3] * 6`, which can not be 0. Similarly, for the dimension before the expanded dimension (dim 1), if it had `true` contiguity, then its stride must be `stride[2] * 10`, which is always zero.

**Comment by naoyam** [comment]
> @naoyam For example, if you have a contiguous tensor of shape `(4, 5, 1, 6)`, then the stride could be `(30, 6, whatever, 1)`, but if you expand this tensor to `(4, 5, 10, 6)`, the stride has to be `(30, 6, 0, 1)`. And by definition, if we have `true` contiguity at the expanded dimension (dim 2), its stride must be `stride[3] * 6`, which can not be 0. Similarly, for the dimension before the expanded dimension (dim 1), if it had `true` contiguity, then its stride must be `stride[2] * 10`, which is always zero.

Thanks for the explanation. Can you please add this as a comment to the code? Maybe at `TensorDomain::getContiguousContiguity`?

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 2 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:1294]: Just trying to save an instance creation; I don't think it will make much difference here. And yes, we should also use zero val, because why not.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 1 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 3 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp]: fixed

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 4 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:1342]: added a comment above describing what it does

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 5 above)
Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:1356]: Isn't this redundant with the above code?
```C++
    if (is_expanded) {
      extent = FusionGuard::getCurFusion()->oneVal();
      shape_extent = &expanded_extent;
    }
```

**Comment by zasdfgbnm** [comment]
@naoyam I have resolved all review comments

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:1356]: OK, then that seems fine. How about an expanded domain with no shape setting? We don't allow symbolic expanded extents, do we?

Comment (zasdfgbnm) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:1356]: We do support symbolic expanded extents. I added a commit to this PR https://github.com/csarofeen/pytorch/pull/2049/commits/3bacbac57ef610efee60ab741a5f940fe16b625a to switch my test to symbolic expanded extents. However, the shape inference in the executor seems to not support input tensors with expanded extents well, so I have to manually make the extent an input of the fusion to make it work. We can fix this issue in a followup PR.

**Review by naoyam — APPROVED** [review]
LGTM.

File: torch/csrc/jit/codegen/cuda/tensor_view.cpp (hunk 5 above)
Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/tensor_view.cpp:1356]: OK, didn't know that. Approved the PR.
